core/shared/src/main/scala/zio/ZEnvironment.scala

```diff
@@ -308,25 +308,27 @@ object ZEnvironment {
      * Constructs a patch that describes the updates necessary to transform the
      * specified old environment into the specified new environment.
      */
-    def diff[In, Out](oldValue: ZEnvironment[In], newValue: ZEnvironment[Out]): Patch[In, Out] = {
-      val sorted = newValue.map.toList.sortBy { case (_, (_, index)) => index }
-      val (missingServices, patch) = sorted.foldLeft[(Map[LightTypeTag, (Any, Int)], Patch[In, Out])](
-        oldValue.map -> Patch.Empty().asInstanceOf[Patch[In, Out]]
-      ) { case ((map, patch), (tag, (newService, newIndex))) =>
-        map.get(tag) match {
-          case Some((oldService, oldIndex)) =>
-            if (oldService == newService && oldIndex == newIndex)
-              map - tag -> patch
-            else
+    def diff[In, Out](oldValue: ZEnvironment[In], newValue: ZEnvironment[Out]): Patch[In, Out] =
+      if (oldValue == newValue) Patch.Empty().asInstanceOf[Patch[In, Out]]
+      else {
+        val sorted = newValue.map.toList.sortBy { case (_, (_, index)) => index }
+        val (missingServices, patch) = sorted.foldLeft[(Map[LightTypeTag, (Any, Int)], Patch[In, Out])](
+          oldValue.map -> Patch.Empty().asInstanceOf[Patch[In, Out]]
+        ) { case ((map, patch), (tag, (newService, newIndex))) =>
+          map.get(tag) match {
+            case Some((oldService, oldIndex)) =>
+              if (oldService == newService && oldIndex == newIndex)
+                map - tag -> patch
+              else
+                map - tag -> patch.combine(AddService(newService, tag))
+            case _ =>
               map - tag -> patch.combine(AddService(newService, tag))
-          case _ =>
-            map - tag -> patch.combine(AddService(newService, tag))
+          }
+        }
+        missingServices.foldLeft(patch) { case (patch, (tag, _)) =>
```

Comment:
There's unfortunately still massive allocation in here. I think we can do a few things:

1. Short term, make sure the happy path (no differences) is special-cased and involves no allocations (actually, that's impossible, since you can't even look at the contents of a map without allocations, but we could get close) and minimal indirection;
2. Long-term, see about NOT using Scala's Map for the environment, but rather switch to a register + array-based implementation that is ideal for "small maps", and could achieve minimal allocations for many operations.

Reply:
The first one is what we're doing here. If the two maps are the same just immediately return an empty patch. The second one we could definitely look at. I think the size we have to optimize for is approximately one hundred services so we just have to make sure we actually come out ahead there.